### DISCUSSION TITLE
AdvancedBlocking

### DISCUSSION THREAD

**Discussion by PeterKnotek** [TechnitiumSoftware/DnsServer#970] [General]
HI

I have a little problem with Adv.Block
Here is my cfg 

{
  "enableBlocking": true,
  "blockListUrlUpdateIntervalHours": 24,
  "localEndPointGroupMap": {},
  "networkGroupMap": {
    "10.15.100.160": "NetflixAllowed",
    "10.15.19.70": "NetflixAllowed",
    "172.16.0.0/12": "fully_blocked",
    "10.0.0.0/8": "fully_blocked",
    "172.23.0.0/16": "Printers",
    "10.15.20.39": "NoESET",
    "172.29.1.11": "NoESET"
                                    },
  "groups": [
   {
      "name": "NoESET",
      "enableBlocking": true,
      "allowTxtBlockingReport": true,
      "blockAsNxDomain": true,
      "blockingAddresses": [
        "127.0.0.1"
      ],
      "allowed": [],
      "blocked": [],
      "allowListUrls": [],
      "blockListUrls": [],
      "allowedRegex": [
        "^(.*\\.)?google\\..*$",
        "^(.*\\.)?unlp\\.sk$",
        "^(.*\\.)?unlp\\.local$",
        "^www\\.vszp\\.sk$",
        "^www\\.union\\.sk$",
        "^www\\.dovera\\.sk$",
        "^(.*\\.)?data\\.microsoft\\.com$",
        "^(.*\\.)?edge\\.microsoft\\.com$",
        "^(.*\\.)?mp\\.microsoft\\.com$",
        "^(.*\\.)?download\\.microsoft\\.com$",
        "^(.*\\.)?go\\.microsoft\\.com$",
        "^(.*\\.)?download\\.windowsupdate\\.net$",
        "^(.*\\.)?download\\.windowsupdate\\.com$",
        "^(.*\\.)?ctldl\\.windowsupdate\\.com$",
        "^(.*\\.)?cxcs\\.microsoft\\.net$",
        "^(.*\\.)?ecs\\.office\\.com$",
        "^(.*\\.)?update\\.microsoft\\.com$",
        "^(.*\\.)?config\\.edge\\.skype\\.com$",
        "^(.*\\.)?api\\.msn\\.com$",
        "^(.*\\.)?assets\\.msn\\.com$",
        "^(.*\\.)?login\\.live\\.com$",
        "^(.*\\.)?digicert\\.com$",
        "^(.*\\.)?azureedge\\.net$",
        "^(.*\\.)?login\\.microsoftonline\\.com$",
        "^(.*\\.)?wdcp\\.microsoft\\.com$",
        "^(.*\\.)?wdcpalt\\.microsoft\\.com$",
        "^(.*\\.)?akamaized\\.net$",
        "^(.*\\.)?windows\\.msn\\.com$",
        "^(.*\\.)?trafficmanager\\.net$",
        "^(.*\\.)?cloudapp\\.azure\\.com$",
        "^(.*\\.)?content\\.office\\.net$",
        "^(.*\\.)?support\\.microsoft\\.com$",
        "^(.*\\.)?smartscreen\\.microsoft\\.com$",
        "^(.*\\.)?msauth\\.net$",
        "^(.*\\.)?nczisk\\.sk$",
        "^(.*\\.)?www\\.bing\\.com$",
        "^(.*\\.)?storage-vszp-prod\\.predu\\.sk$",
        "^(.*\\.)?dns\\.msftncsi\\.com$",
        "^(.*\\.)?eset\\..*$"
      ],
      "blockedRegex": [
        ".*"
      ],
      "regexAllowListUrls": [],
      "regexBlockListUrls": [],
      "adblockListUrls": []
    },
    {
      "name": "NetflixAllowed",
      "enableBlocking": true,
      "allowTxtBlockingReport": true,
      "blockAsNxDomain": true,
      "blockingAddresses": [],
      "allowed": [
        "netflix.com",
        "*.netflix.com"
      ],
      "blocked": [],
      "allowListUrls": [],
      "blockListUrls": [],
      "allowedRegex": [],
      "blockedRegex": [],
      "regexAllowListUrls": [],
      "regexBlockListUrls": [],
      "adblockListUrls": []
    },
    {
      "name": "fully_blocked",
      "enableBlocking": true,
      "allowTxtBlockingReport": true,
      "blockAsNxDomain": true,
      "blockingAddresses": [
        "127.0.0.1"
      ],
      "allowed": [],
      "blocked": [],
      "allowListUrls": [],
      "blockListUrls": [],
      "allowedRegex": [],
      "blockedRegex": [
        "^(.*\\.)?sme\\..*$",
        "^(.*\\.)?alibaba\\..*$",
        "^(.*\\.)?aliexpress\\..*$",
        "^(.*\\.)?jwplayer\\..*$",
        "^(.*\\.)?netflix\\..*$",
        "^(.*\\.)?ccleaner\\..*$",
        "^(.*\\.)?avast\\..*$",
        "^(.*\\.)?tahaj\\..*$",
        "^(.*\\.)?telegram\\..*$",
        "^(.*\\.)?kukaj\\..*$",
        "^(.*\\.)?uloz\\..*$",
        "^(.*\\.)?sledujteto\\..*$",
        "^(.*\\.)?streamuj\\..*$",
        "^(.*\\.)?fastshare\\..*$",
        "^(.*\\.)?prehraj\\..*$",
        "^(.*\\.)?datoid\\..*$",
        "^(.*\\.)?sdilej\\..*$",
        "^(.*\\.)?zalohuj\\..*$",
        "^(.*\\.)?webshare\\..*$",
        "^(.*\\.)?uloz\\..*$",
        "^(.*\\.)?flixtor\\..*$",
        "^(.*\\.)?kra\\..*$",
        "^(.*\\.)?mcafee\\..*$",
        "^(.*\\.)?horoskopy\\.cz$",
        "^(.*\\.)?disqus\\.com$",
        "^(.*\\.)?blesk\\.cz$",
        "^(.*\\.)?t\\.me$",
        "^(.*\\.)?mega\\.io$"
      ],
      "regexAllowListUrls": [],
      "regexBlockListUrls": [],
      "adblockListUrls": []
    },
    {
      "name": "Printers",
      "enableBlocking": true,
      "allowTxtBlockingReport": true,
      "blockAsNxDomain": true,
      "blockingAddresses": [
        "127.0.0.1"
      ],
      "allowed": [
        "ccc.hpeprint.com",
        "h10141.www1.hp.com"
      ],
      "blocked": [],
      "allowListUrls": [],
      "blockListUrls": [],
      "allowedRegex": [],
      "blockedRegex": [
        ".*"
      ],
      "regexAllowListUrls": [],
      "regexBlockListUrls": [],
      "adblockListUrls": []
    }
  ]
}

My problem is : 
  "10.15.100.160": "NetflixAllowed",
  "10.15.19.70": "NetflixAllowed",
  -  why IP 10.15.100.161 has DNS ansver form eg kukaj.io ( is blocked regex ##  "^(.*\\.)?kukaj\\..*$", ##
     - ich change IP to 10.15.100.161 block for kukaj.io work 
     Need block for all some domain eg google | netflix | and for other IP allow allow netflix pr google or other domains
   But ... not work for IP 
  "10.15.100.160": "NetflixAllowed",
  "10.15.19.70": "NetflixAllowed",

Anyone help?
Thanks

**Comment by ShreyasZare** [comment]
Thanks for the post. I am not exactly sure I understood your query since you have mentioned `10.15.100.161` twice and said it both works and not work.

Also, how are you testing the config? The test must be done from the same client IP address using tools like `nslookup`.

**Reply by PeterKnotek** [reply]
HI 

Soory - My fault ... 

If have ANY IP in IP group  "NetflixAllowed" ( eg 10.15.100.160 - or 161 or 192.168.5.50 )  then not aplied rulles in 
{
"name": "fully_blocked",
"enableBlocking": true,
"allowTxtBlockingReport": true,
"blockAsNxDomain": true,
"blockingAddresses": [
"127.0.0.1"
],
"allowed": [],
"blocked": [],
"allowListUrls": [],
"blockListUrls": [],
"allowedRegex": [],
"blockedRegex": [
"^(.\.)?sme\..$",
"^(.\.)?alibaba\..$",
"^(.\.)?aliexpress\..$",
"^(.\.)?jwplayer\..$",
"^(.\.)?netflix\..$",
"^(.\.)?ccleaner\..$",
"^(.\.)?avast\..$",
"^(.\.)?tahaj\..$",
"^(.\.)?telegram\..$",
"^(.\.)?kukaj\..$",
"^(.\.)?uloz\..$",
"^(.\.)?sledujteto\..$",
"^(.\.)?streamuj\..$",
"^(.\.)?fastshare\..$",
"^(.\.)?prehraj\..$",
"^(.\.)?datoid\..$",
"^(.\.)?sdilej\..$",
"^(.\.)?zalohuj\..$",
"^(.\.)?webshare\..$",
"^(.\.)?uloz\..$",
"^(.\.)?flixtor\..$",
"^(.\.)?kra\..$",
"^(.\.)?mcafee\..$",
"^(.\.)?horoskopy\.cz$",
"^(.\.)?disqus\.com$",
"^(.\.)?blesk\.cz$",
"^(.\.)?t\.me$",
"^(.\.)?mega\.io$"
],
"regexAllowListUrls": [],
"regexBlockListUrls": [],
"adblockListUrls": []
},

I try to combine the rules - but still the same result
nslookup kukaj.io or kukaj.to - send back result 

monitor is device with IP 10.15.100.160
root@monitor:~# nslookup kukaj.io
Server:         172.30.0.10
Address:        172.30.0.10#53

Non-authoritative answer:
Name:   kukaj.io
Address: 172.67.219.131
Name:   kukaj.io
Address: 104.21.45.221
Name:   kukaj.io
Address: 2606:4700:3032::6815:2ddd
Name:   kukaj.io
Address: 2606:4700:3035::ac43:db83


admin is device with IP 172.30.1.200
root@unlp-admin:~# nslookup sme.sk
;; Got recursion not available from 172.30.0.10, trying next server
Server:         172.30.0.20
Address:        172.30.0.20#53

** server can't find sme.sk: NXDOMAIN

Thanks

**Reply by ShreyasZare** [reply]
> If have ANY IP in IP group "NetflixAllowed" ( eg 10.15.100.160 - or 161 or 192.168.5.50 ) then not aplied rulles in

The network to group map will map the closest network address to the client's IP address and select the group. Only one group will be selected. So, in your config, the closest network is for the IP address itself which will always be selected over `10.0.0.0/8`. Combining two groups is not possible. You will need to define the groups keeping this is mind.

**Reply by PeterKnotek** [reply]
Thanks 

I do not understand that
But can you give a small example ??

Thanks

**Reply by ShreyasZare** [reply]
You can create a local file and store all these regex in it and then use the file path with the `regexBlockListUrls` array in the config with multiple groups so that you have one single file that is reusable with all the groups that need it. 

The local file path should be in URL format, for example: `file:///home/folder/myregexblocklist.txt` if you are on Linux or `file:///c:/folder/myblocklist.txt` if you are on Windows.

**Reply by PeterKnotek** [reply]
Hi - thanks for answer 

But primary problem or question is .... for example
Have ... 
"172.16.0.0/12": "ClassAB",
"10.0.0.0/8": "CLassAB", 
"10.15.100.150": "ExceptionIP_1"

And 
For CLassAB in one Rules block with regex expresion
expample
"^(..)?horoskopy.cz$",
"^(..)?disqus.com$",
"^(..)?blesk.cz$",
"^(..)?t.me$",
"^(..)?mega.io$" 
 "^(. \.)?netflix\.. $", 
 
 And in second rules  for  "ExceptionIP_1"
 - Allow only eg netflix domain, however, simultaneously block domains listed in the rule for ClassAB.
 
 Is it even possible within the configuration?
 If yes, then how can it be achieved? Specifically, where can I find a manual for AdvancedBlocking?
 
Imagine a scenario where a packet traverses a tree of rules on a firewall with the QUICK match feature. Once a match is found, it doesn’t proceed further. In other words, allow-listed websites are checked first. For example, if an IP address (e.g., 10.15.100.150) queries netflix.com and matches a rule, the DNS query stops there. Similarly, if a request comes from 10.15.100.160 to kukaj.io and matches a ClassAB rule (resulting in blocking), it’s discarded. However, if a request from 10.15.100.160 goes to a site like antik.sk and isn’t globally blocked or blocked via Advanced Blocking rules, it responds.”

Soory for my bad UK langue

Thanks for your time

**Reply by ShreyasZare** [reply]
In you example, you need to add those same regex rules in your `ExceptionIP_1` too. So basically, both the `ClassAB` and `ExceptionIP_1` groups should have the same regex blocking rules. Then add your domain name to exclude only in `ExceptionIP_1` group.

**Reply by PeterKnotek** [reply]
That not work 
If have IP 10.15.100.150 - the rule for blocking other sites does not apply to IP 10.15.100.150

**Reply by ShreyasZare** [reply]
> If have IP 10.15.100.150 - the rule for blocking other sites does not apply to IP 10.15.100.150

You have to copy the same regex in to the group that is used for that IP. You need to copy paste the blocking regex in all the groups.

**Reply by PeterKnotek** [reply]
> > Ak máte IP 10.15.100.150 - pravidlo pre blokovanie iných stránok neplatí pre IP 10.15.100.150
> 
> Musíte skopírovať rovnaký regulárny výraz do skupiny, ktorá sa používa pre danú adresu IP. Musíte skopírovať a vložiť blokujúci regulárny výraz do všetkých skupín.

A small example PLEASE - I still don't understand

**Reply by ShreyasZare** [reply]
Here is a small example where the same regex is configured in both the groups and one group has additional allowed domain name for netflix. Just to keep the example small, I am adding just one regex entry so you need to add the remaining regex entries.
```
{
	"enableBlocking": true,
	"blockListUrlUpdateIntervalHours": 24,
	"localEndPointGroupMap": {},
	"networkGroupMap": {
		"10.15.100.160": "NetflixAllowed",
		"10.0.0.0/8": "fully_blocked",
	},
	"groups": [
		{
			"name": "NetflixAllowed",
			"enableBlocking": true,
			"allowTxtBlockingReport": true,
			"blockAsNxDomain": true,
			"blockingAddresses": [],
			"allowed": [
				"netflix.com"
			],
			"blocked": [],
			"allowListUrls": [],
			"blockListUrls": [],
			"allowedRegex": [],
			"blockedRegex": [
				".\\.kukaj\\..$"
			],
			"regexAllowListUrls": [],
			"regexBlockListUrls": [],
			"adblockListUrls": []
		},
		{
			"name": "fully_blocked",
			"enableBlocking": true,
			"allowTxtBlockingReport": true,
			"blockAsNxDomain": true,
			"blockingAddresses": [
				"127.0.0.1"
			],
			"allowed": [],
			"blocked": [],
			"allowListUrls": [],
			"blockListUrls": [],
			"allowedRegex": [],
			"blockedRegex": [
				".\\.kukaj\\..$"
			],
			"regexAllowListUrls": [],
			"regexBlockListUrls": [],
			"adblockListUrls": []
		}
	]
}
```

Note that you need to follow the regex described in [here](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-language-quick-reference).

Also note that since the regex is being used inside JSON format strings, you need to escape the backslash character so that the JSON format is correct.

**Reply by PeterKnotek** [reply]
Thanks very very very - works

**Reply by ShreyasZare** [reply]
You're welcome.

**Reply by PeterKnotek** [reply]
HI

One question :
```
{
      "name": "AA",
      "enableBlocking": true,
      "allowTxtBlockingReport": true,
      "blockAsNxDomain": true,
      "blockingAddresses": [ "127.0.0.1" ],
      "allowed": [],
      "blocked": [],
      "allowListUrls": [],
      "blockListUrls": [],
      "allowedRegex": [],
      "regexAllowListUrls": [ 
                                       "http://172.1.1.10/TechnitiumDomain/XXX"
                                       ],
      "blockedRegex": [
                                 ".*"
                                 ],
      "regexBlockListUrls": [],
      "adblockListUrls": []
    },
```
but when the content of the page is like this... it doesn't work
"^(..)?sme..$"
"^(..)?alibaba..$"
"^(..)?aliexpress..$"
"^(..)?jwplayer..$"
"^(..)?netflix..$"
"^(..)?ccleaner..$"
"^(..)?avast..$"
"^(..)?tahaj..$"
"^(..)?telegram..$"
"^(..)?kukaj..$"
"^(..)?uloz..$"
"^(..)?sledujteto..$"
"^(..)?streamuj..$"

Pleas HELP ME
And last .... 
how do I refresh the content???

Thanks

**Reply by ShreyasZare** [reply]
Thanks for the details. Does the regex text file include the regex with those double quotes? If yes, then remove those quotes.

Since you have the regex list configured as allowed list, and that there is nothing configured for blocking, there wont be anything to do for the app here. All those domain names in the list are already allowed and would resolve.

**Reply by PeterKnotek** [reply]
Thanks very very

"^(..)?streamuj..$"  - not work 
^(..)?streamuj..$   - not work
(..)?streamuj..$      - work
but only if this value is like web content

Last question
in cfg exist :  "blockListUrlUpdateIntervalHours": 1,
can use     :    "allowListUrlsUpdateIntervalHours":1, ????
 
Thanks

**Reply by ShreyasZare** [reply]
You're welcome. The same block list update interval option is used for updating the allow lists too.

**Reply by PeterKnotek** [reply]
Thanks

Bud auto cerate every one hour file : 
/etc/dns/apps/Advanced Blocking/blocklists/4cbc8e0d8a6b017fb2d54ecfe0831162549cdce334f30e36820764c971e3b998
- I think this file generate by defaul CFG
  - 
But also files where generated by maunal CFG with : "regexAllowListUrls" or "allowListUrls" is created in " /etc/dns/apps/Advanced Blocking/blocklists/ "
- I thing ... when is fille ( Allow list ) - the folder should be "/etc/dns/apps/Advanced Blocking/allowlist" - not blocklists

And last idea : can be create cfg for : allowListUrlsUpdateIntervalHour or allowListUrlsUpdateIntervalMinutes ???
- eg minicron ??

And this is probably the error or bug : 
1 -
- if change primary www contect where is data for "allowListUrls"
  -  and confirm APPLY in the APP setting for AdvancedBlocking
    - then NOT generate file whit fresh content - content is old // floods
    - the file will not be overwritten
2 -
- when delete files in "/etc/dns/apps/Advanced Blocking/blocklists/*" 
- if change primary www contect where is data for "allowListUrls"
  -  and confirm APPLY in the APP setting for AdvancedBlocking
    - then generate file whit fresh content - content is OK
        

Thanks very very very much for your answer

**Reply by ShreyasZare** [reply]
> Bud auto cerate every one hour file : /etc/dns/apps/Advanced Blocking/blocklists/4cbc8e0d8a6b017fb2d54ecfe0831162549cdce334f30e36820764c971e3b998

Those are cache files so that the URL can be loaded without fetching it every time from the Internet. As a user, you do not need to change any of these file on the disk. Always use the GUI to do the config.
 
> But also files where generated by maunal CFG with : "regexAllowListUrls" or "allowListUrls" is created in " /etc/dns/apps/Advanced Blocking/blocklists/ "
> 
>   I thing ... when is fille ( Allow list ) - the folder should be "/etc/dns/apps/Advanced Blocking/allowlist" - not blocklists

That is just a cache folder and name of the folder does not matter. Both block list and allow list cache files are stored there.

> And last idea : can be create cfg for : allowListUrlsUpdateIntervalHour or allowListUrlsUpdateIntervalMinutes ???
> 
>     * eg minicron ??

These lists are not frequently updates so an hour based interval is most suitable for this which is why there is no option to configure it in minutes. People configuring 1 min update interval would cause these list webservers to get flooded unnecessarily. 

There is no plan to have a separate interval variable for allow list since it does not make any difference to treat it separately from block list.


> And this is probably the error or bug : 1 -
> 
>     * if change primary www contect where is data for "allowListUrls"
>       
>       * and confirm APPLY in the APP setting for AdvancedBlocking
>       * then NOT generate file whit fresh content - content is old // floods
>       * the file will not be overwritten
>         2 -

The file update mechanism update files based on the time interval and the cache file's date last modified time stamp. It will not always fetch the same file if not needed. This is not a bug but optimization done to prevent frequently fetching lists.

>     * when delete files in "/etc/dns/apps/Advanced Blocking/blocklists/*"
> 
>     * if change primary www contect where is data for "allowListUrls"
>       
>       * and confirm APPLY in the APP setting for AdvancedBlocking
>       * then generate file whit fresh content - content is OK

If you delete cache file then it will get fetched as expected. 


> Thanks very very very much for your answer

You're welcome.
